vitessio/vitess
# VReplication: use PK equivalent (PKE) if no PK found

## Description

This allows VReplication to be more MySQL compatible as MySQL will automatically try to use a PRIMARY KEY equivalent (PKE) when the ability to uniquely identify rows for internal operations is needed, if no defined PRIMARY KEY exists. It also ensures that the query used on the source and target tablets is more efficient when there is no explicitly defined PRIMARY KEY on the table.

ℹ️ **NOTE:** This modifies/improves the query used on tablets. We can make the same basic improvement for the VDiff query used on source and target, but I'm punting on that for now as [we are in the process of moving VDiff work to tablets](https://github.com/vitessio/vitess/issues/10134) (currently it's done in `vtctl`) which would then allow us to improve the query in the same way this PR does (w/o having to add RPCs for getting the PKEs and use those from `vtctl` -> `vttablet` -> `mysqld`).

## Manual Test

This cut and paste test shows that the VReplication query (`ORDER BY`) matches [what we see in the new unit test:](https://github.com/vitessio/vitess/pull/10192/files#diff-f61d8ccd19060e84eaa9ab08e07c174de6bd8ad667ad1f0b0f0a2397db9d8292R102-R153)

```
$ make build && pushd ./examples/local; ./101_initial_cluster.sh; ./201_customer_tablets.sh; popd; \
mysql -u root --socket=/vt/vtdataroot/vt_0000000100/mysql.sock -e 'set global general_log=ON; use vt_commerce; CREATE TABLE onepke_t (id INT NOT NULL, col1 VARCHAR(25), UNIQUE KEY (id)); CREATE TABLE onemcpke_t (col1 VARCHAR(25) NOT NULL, col2 VARCHAR(25) NOT NULL, col3 VARCHAR(25) NOT NULL, col4 VARCHAR(25), UNIQUE KEY c4_c2_c1 (col4, col2, col1), UNIQUE KEY c1_c2 (col1, col2), UNIQUE KEY c1_c2_c4 (col1, col2, col4), KEY nc1_nc2 (col1, col2)); CREATE TABLE twomcpke_t (col1 VARCHAR(25) NOT NULL, col2 VARCHAR(25) NOT NULL, col3 VARCHAR(25) NOT NULL, col4 VARCHAR(25), UNIQUE KEY (col4), UNIQUE KEY c4_c2_c1 (col4, col2, col1), UNIQUE KEY c1_c2_c3 (col1, col2, col3), UNIQUE KEY c1_c2 (col1, col2)); CREATE TABLE oneintpke1charpke_t (col1 VARCHAR(25) NOT NULL, col2 VARCHAR(25) NOT NULL, col3 VARCHAR(25) NOT NULL, id1 INT NOT NULL, id2 INT NOT NULL, UNIQUE KEY c1_c2 (col1, col2), UNIQUE KEY id1_id2 (id1, id2)); CREATE TABLE twointvsvcharpke_t (col1 VARCHAR(25) NOT NULL, id1 INT NOT NULL, id2 INT NOT NULL, UNIQUE KEY c1 (col1), UNIQUE KEY id1_id2 (id1, id2)); CREATE TABLE tinyintvsbigint_t (tid1 TINYINT NOT NULL, id1 INT NOT NULL, UNIQUE KEY tid1 (tid1), UNIQUE KEY id1 (id1)); CREATE TABLE vcharintvsinttwovchar_t (id1 INT NOT NULL, col1 VARCHAR(25) NOT NULL, col2 VARCHAR(25) NOT NULL, UNIQUE KEY col1_id1 (col1, id1), UNIQUE KEY id1_col1_col2 (id1, col1, col2)); CREATE TABLE vcharvsintthree_t (id1 INT NOT NULL, id2 INT NOT NULL, id3 INT NOT NULL, col1 VARCHAR(50) NOT NULL, UNIQUE KEY col1 (col1), UNIQUE KEY id1_id2_id3 (id1, id2, id3))' ; \
vtctlclient --server=localhost:15999 MoveTables -- --source commerce --all --tablet_types=primary Create customer.commerce2customer ; \
sleep 10 && grep "_t order by" /vt/vtdataroot/vt_0000000100/data/ip-172-31-41-160.log
```

Results being:
```
2022-05-09T14:56:24.473476Z	   87 Query	select col1, col2, col3, id1, id2 from oneintpke1charpke_t order by id1, id2
2022-05-09T14:56:25.553185Z	   95 Query	select col1, col2, col3, col4 from onemcpke_t order by col1, col2
2022-05-09T14:56:26.590831Z	  103 Query	select id, col1 from onepke_t order by id
2022-05-09T14:56:28.652515Z	  118 Query	select tid1, id1 from tinyintvsbigint_t order by tid1
2022-05-09T14:56:29.685208Z	  126 Query	select col1, id1, id2 from twointvsvcharpke_t order by id1, id2
2022-05-09T14:56:30.734427Z	  134 Query	select col1, col2, col3, col4 from twomcpke_t order by col1, col2
2022-05-09T14:56:31.766876Z	  142 Query	select id1, col1, col2 from vcharintvsinttwovchar_t order by col1, id1
2022-05-09T14:56:32.803098Z	  150 Query	select id1, id2, id3, col1 from vcharvsintthree_t order by id1, id2, id3
```

## Related Issue(s)
  - Fixes: https://github.com/vitessio/vitess/issues/7120
  
## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Documentation: https://github.com/vitessio/website/pull/1021
